### DISCUSSION TITLE
SEO optimization and automatization of OpenCart 4

### DISCUSSION THREAD

**Discussion by ppalashturov** [opencart/opencart#12930] [Ideas]
Hello,
We from OpenCart Bulgaria have an idea to do some extra SEO optimization and automatization of OpenCart 4. We want to apply and make a pull request with the following functions:

1. Adding a PHP library for the SEO tags and adding the Documents library to add the tags (meta, schema, etc.) to each page. We are planning to use: https://github.com/melbahja/seo

2. Add functionality to generate sitemap.xml.

3. Add functionality for Image optimization like selecting image formats jpg or WebP from the admin panel when caching images. Add LazyLoad library.

5. Editable title or H1 tags in the pages of the default template.

6. Adding functionality for automatic generation of URL addresses.

As well as several other things for the better SEO.

Best regards!

**Comment by ppalashturov** [comment]
@danielkerr  What do you think of the idea?

**Comment by pemaier** [comment]
In case of automatization ... you have an idea how to handle upgrades of big amount (several thousands) of categories, products? In version 3 it was not required to enter an SEO entry for each language and for each store. So what is you suggestion how to handle upgrades because in version 4 it is required?
